### PR TITLE
Support for signed Releases file

### PR DESCRIPTION
**Pull request by dig412**
This PR adds support for generating and signing the Releases file using a GPG key.

I wrote this because:
 - `apt` complains about pulling from unsigned repositories, and will refuse to do so unless you add `trusted=yes` to the sources file. Not a major hardship, but hey. 
 - I'm enjoying writing Go and I wanted to see how difficult it would be. Turns out not very - the `openpgp` library does all the heavy lifting.

I've added easy key generation to the CLI as I dislike faffing around with GPG keyrings. Signing is configurable via `conf.json`, and if it's disabled the Releases file won't be created.

This adds a fair bit more code and a large-ish dependency, so I understand if you don't want to merge it to mainline in order to keep things simple.

### CODE REVIEW

**Comment by esell** [comment]
hey @dig412, thanks for another PR! let me try and grok this (i'm sort of a n00b as well) and get back to you. i certainly see the benefit of this and it could make it more of a production type app.

File: main.go [hunk 1]
```diff
@@ -56,8 +62,17 @@ var (
...
+	keyName      = flag.String("kn", "", "Name for the siging key")
+	keyEmail     = flag.String("ke", "", "Email address")
 	parsedconfig = conf{}
 	mywatcher    *fsnotify.Watcher
+
+
+	//Create a package level time function so we can mock it out
+	Now = func() time.Time {
+		return time.Now()
+	}
```
Comment (esell) [inline main.go]: Curious about this guy, why not just use the `time` package? It appears in the tests the only thing that changes is that a set value is returned.

File: main.go [hunk 2]
```diff
@@ -98,6 +113,21 @@ func main() {
 		os.Exit(0)
 	}
 
+	if *generateSigningKey {
+
+		workingDirectory, err := os.Getwd()
+		if err != nil {
+			fmt.Errorf("Unable to get current working directory: %s", err)
```
Comment (esell) [inline main.go]: would it make sense to use `log.Fatal` here just to keep things consistent?

File: main.go [hunk 3]
```diff
@@ -302,15 +344,221 @@ func createPackagesGz(config conf, distro, section, arch string) error {
...
+    		relPath, err := filepath.Rel(workingDirectory, path)
+    		spath := filepath.ToSlash(relPath)
+    		f, err := os.Open(path)
+    		var (
+    			md5hash    = md5.New()
+    			sha1hash   = sha1.New()
+    			sha256hash = sha256.New()
+    		)
+    		if _, err = io.Copy(io.MultiWriter(md5hash, sha1hash, sha256hash), f); err != nil {
+    			log.Println("error with the md5 hash: ", err)
```
Comment (esell) [inline main.go]: should this error stop everything? is it important enough to return here? i'm super foggy on the package stuff so i'll defer to you :)

File: main.go [hunk 4]
```diff
@@ -302,15 +344,221 @@ func createPackagesGz(config conf, distro, section, arch string) error {
...
+    		}
+    		fmt.Fprintf(&md5Sums, " %s %d %s\n",
+    			hex.EncodeToString(md5hash.Sum(nil)),
+    			file.Size(), spath)
+
+    		if _, err = io.Copy(sha1hash, f); err != nil {
+    			log.Println("error with the sha1 hash: ", err)
```
Comment (esell) [inline main.go]: should this error stop everything? is it important enough to return here? i'm super foggy on the package stuff so i'll defer to you :)

File: main.go [hunk 5]
```diff
@@ -302,15 +344,221 @@ func createPackagesGz(config conf, distro, section, arch string) error {
...
+    		}
+    		fmt.Fprintf(&sha1Sums, " %s %d %s\n",
+    			hex.EncodeToString(sha1hash.Sum(nil)),
+    			file.Size(), spath)
+
+    		if _, err = io.Copy(sha256hash, f); err != nil {
+    			log.Println("error with the sha256 hash: ", err)
```
Comment (esell) [inline main.go]: should this error stop everything? is it important enough to return here? i'm super foggy on the package stuff so i'll defer to you :)

File: main.go [hunk 6]
```diff
@@ -302,15 +344,221 @@ func createPackagesGz(config conf, distro, section, arch string) error {
...
+
+	outfile.WriteString("MD5Sum:\n")
+	outfile.WriteString(md5Sums.String())
+	outfile.WriteString("SHA1:\n")
+	outfile.WriteString(sha1Sums.String())
+	outfile.WriteString("SHA256:\n")
+	outfile.WriteString(sha256Sums.String())
+
+	if config.EnableSigning {
+		signRelease(config, outfile.Name())
```
Comment (esell) [inline main.go]: should this be checked to see if an error is returned?

File: main.go [hunk 7]
```diff
@@ -113,14 +143,17 @@ func main() {
...
 					mutex.Lock()
-					if filepath.Base(event.Name) != "Packages.gz" {
+					if filepath.Ext(event.Name) == ".deb" {
 						log.Println("Event: ", event)
 						distroArch := destructPath(event.Name)
 						if err := createPackagesGz(parsedconfig, distroArch[0], distroArch[1], distroArch[2]); err != nil {
 							log.Printf("error creating package: %s", err)
 						}
+						if parsedconfig.EnableSigning {
+							createRelease(parsedconfig, distroArch[0])
```
Comment (esell) [inline main.go]: log returned error here?

**Comment by esell** [comment]
@dig412, i added some questions/suggestions so let me know what you think. most of them are simple things, nothing major.

File: main.go (hunk 1 above)
Comment (dig412) [reply main.go]: That's the reason - the current time is written to the Release file so I can't compare it to a known good value without fixing the time, so wrapping time.Now() in an anonymous function allows it to be overridden.

I found this snipped on Stack Overflow, and it seemed a lot easier than importing a full time mocking library.

File: main.go (hunk 2 above)
Comment (dig412) [reply main.go]: Sure. I'm not sure on the difference between `log` and `fmt` - I'll have another look.

File: main.go (hunk 3 above)
Comment (dig412) [reply main.go]: I guess so - I can't really imagine what kind of errors would come out of that process, but without the hashes the Release file certainly won't work.

**Comment by dig412** [comment]
Thanks @esell, I've pushed another commit that fixes the `err` / `log` / `fmt` changes you suggested.

**Comment by dig412** [comment]
You're welcome - thanks for writing this really useful bit of software. I'm glad to be able to give back.

**Comment by esell** [comment]
eek, looks like the test broke, i'll take a look and see if i can track it down...

**Pull request merged**